cmd/snap-bootstrap/triggerwatch/evdev.go

```diff
@@ -0,0 +1,158 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package triggerwatch
+
+import (
+	"fmt"
+
+	// TODO:UC20: not packaged, reimplment the minimal things we need?
+	evdev "github.com/gvalkov/golang-evdev"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+type keyEvent struct {
+	Dev triggerDevice
+	Err error
+}
+
+type triggerEventFilter struct {
+	Key string
+}
+
+var (
+	evKeyCapability = evdev.CapabilityType{Type: evdev.EV_KEY, Name: "EV_KEY"}
+
+	// chooserTriggerKey is what activates the chooser when held down
+	chooserTriggerKey = evdev.CapabilityCode{Code: 2, Name: "KEY_1"}
+)
+
+func init() {
+	trigger = &evdevInput{}
+}
+
+type evdevKeyboardInputDevice struct {
+	keyCode uint16
+	dev     *evdev.InputDevice
+}
+
+func (e *evdevKeyboardInputDevice) WaitForTrigger(ch chan keyEvent) {
+	const triggerHeldCount = 20
+
+	heldCount := uint(0)
+
+	// TODO: find a reliable way to get/set key repeat rate? EVIOCGREP seems
+	// to return bad info; the support varies from driver to driver,
+	// gpio-keys may require 'autorepeat' enabled at the devicetree level
+
+	rep := e.dev.GetRepeatRate()
+	// restore original repeat settings
+	defer e.dev.SetRepeatRate(rep[0], rep[1])
+	e.dev.SetRepeatRate(1000, 0)
+
+	for {
+		ies, err := e.dev.Read()
+		if err != nil {
+			ch <- keyEvent{Err: err, Dev: e}
+			break
+		}
+		for _, ie := range ies {
+			if ie.Type != evdev.EV_KEY || ie.Code != e.keyCode {
+				continue
+			}
+			kev := evdev.NewKeyEvent(&ie)
+			switch kev.State {
+			case evdev.KeyHold, evdev.KeyDown:
+				heldCount++
+			case evdev.KeyUp:
+				heldCount = 0
+			}
+			logger.Noticef("%s held: %v", e.dev.Phys, heldCount)
+			if heldCount >= triggerHeldCount {
+				ch <- keyEvent{Dev: e}
+				break
+			}
+		}
+	}
+}
+
+func (e *evdevKeyboardInputDevice) String() string {
+	return fmt.Sprintf("%s: %s", e.dev.Phys, e.dev.Name)
+}
+
+func (e *evdevKeyboardInputDevice) Close() {
+	e.dev.File.Close()
+}
+
+type evdevInput struct{}
+
+func (e *evdevInput) FindMatchingDevices(filter triggerEventFilter) ([]triggerDevice, error) {
+	devices, err := evdev.ListInputDevices()
+	if err != nil {
+		return nil, fmt.Errorf("cannot list input devices: %v", err)
+	}
+
+	// XXX: this is set up to support key input only
+	kc, ok := strToKey[filter.Key]
+	if !ok {
+		return nil, fmt.Errorf("cannot find a key matching the filter %q", filter.Key)
+	}
+	// chooserTriggerKey is what activates the chooser when held down
+	cap := evdev.CapabilityCode{Code: kc, Name: filter.Key}
+
+	match := func(dev *evdev.InputDevice) triggerDevice {
+		for _, cc := range dev.Capabilities[evKeyCapability] {
+			if cc == cap {
+				return &evdevKeyboardInputDevice{
+					dev:     dev,
+					keyCode: uint16(cap.Code),
+				}
+			}
+		}
+		return nil
+	}
+	// Find the first devices that has keys and the trigger key
```

Comment:
this is confusing, we are not finding the first, we are find all of them?  should we find the first? all as now but have a TODO about letting the gadget tell use what to use?

Reply:
Yeah, old comment. Rewritten now.